File: apps/android/app/src/main/java/tech/stygian/presently/oauth/ATProtoOAuthClient.kt
package tech.stygian.presently.oauth

import android.util.Base64
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import org.json.JSONObject
import java.net.URI
import java.net.URLDecoder
import java.net.URLEncoder
import java.security.MessageDigest
import java.security.SecureRandom
import java.time.Instant
import tech.stygian.presently.story.ATProtoBlob
import tech.stygian.presently.story.FlashesActorProfileContract
import tech.stygian.presently.story.FlashesActorProfileFactory
import tech.stygian.presently.story.FlashesStoryContract
import tech.stygian.presently.story.FlashesStoryRecordFactory
import tech.stygian.presently.story.PublishedStory
import tech.stygian.presently.story.StoryPublishingException
import tech.stygian.presently.story.toJson

class ATProtoOAuthClient(
    private val store: SecureOAuthStore,
    private val http: OAuthHttpClient = OAuthHttpClient(),
) {
    private val discovery = OAuthDiscovery(http)
    private val refreshMutex = Mutex()

    suspend fun prepareAuthorization(identifier: String): String {
        val publishedScope = http.getJson(OAuthConfig.ClientId).optString("scope")
            .split(" ")
            .filter(String::isNotBlank)
        if (!publishedScope.containsAll(OAuthConfig.Scopes)) {
            throw OAuthException(
                "Presently’s login service is being updated. Please try again shortly.",
            )
        }
        val result = discovery.discover(identifier)
        val key = DPoPKey.create()
        try {
            val state = randomToken()
            val verifier = randomToken(48)
            val challenge = MessageDigest.getInstance("SHA-256")
                .digest(verifier.toByteArray(Charsets.UTF_8))
                .base64Url()
            val fields = listOf(
                "client_id" to OAuthConfig.ClientId,
                "response_type" to "code",
                "redirect_uri" to OAuthConfig.RedirectUri,
                "scope" to OAuthConfig.Scope,
                "state" to state,
                "code_challenge" to challenge,
                "code_challenge_method" to "S256",
                "login_hint" to result.inputIdentifier,
            )
            val par = dpopFormRequest(result.server.parEndpoint, fields, key, null)
            if (par.first.statusCode !in listOf(200, 201)) {
                throw OAuthException("The authorization server rejected the request.")
            }
            val requestUri = JSONObject(String(par.first.body, Charsets.UTF_8))
                .optString("request_uri")
            if (requestUri.isEmpty()) {
                throw OAuthException("The authorization server returned an invalid PAR response.")
            }
            store.savePending(
                PendingAuthorization(
                    inputIdentifier = result.inputIdentifier,
                    expectedDid = result.did,
                    pdsUrl = result.pdsUrl,
                    issuer = result.server.issuer,
                    authorizationEndpoint = result.server.authorizationEndpoint,
                    tokenEndpoint = result.server.tokenEndpoint,
                    state = state,
                    codeVerifier = verifier,
                    dpopKeyId = key.id,
                    authorizationServerNonce = par.second,
                ),
            )
            return "${result.server.authorizationEndpoint}?" +
                "client_id=${OAuthConfig.ClientId.formEncode()}&" +
                "request_uri=${requestUri.formEncode()}"
        } catch (error: Throwable) {
            key.delete()
            throw error
        }
    }

    suspend fun completeAuthorization(callbackUrl: String): OAuthSession {
        val pending = store.loadPending()
            ?: throw OAuthException("There is no pending OAuth sign-in.")
        val callback = validateCallback(callbackUrl, pending)
        val key = DPoPKey.load(pending.dpopKeyId)
        val result = dpopFormRequest(
            pending.tokenEndpoint,
            listOf(
                "grant_type" to "authorization_code",
                "client_id" to OAuthConfig.ClientId,
                "redirect_uri" to OAuthConfig.RedirectUri,
                "code" to callback,
                "code_verifier" to pending.codeVerifier,
            ),
            key,
            pending.authorizationServerNonce,
        )
        if (result.first.statusCode != 200) {
            throw OAuthException("The authorization server rejected the token request.")
        }
        val token = JSONObject(String(result.first.body, Charsets.UTF_8))
        val subject = token.optString("sub")
        val scopes = token.optString("scope").split(" ").filter(String::isNotBlank)
        val refreshToken = token.optString("refresh_token")
        if (!token.optString("token_type").equals("DPoP", ignoreCase = true) ||
            subject != pending.expectedDid ||
            refreshToken.isEmpty() ||
            !scopes.containsAll(OAuthConfig.Scopes)
        ) {
            throw OAuthException(
                "The token response did not grant the account and permissions Presently requested.",
            )
        }
        discovery.validateAccount(subject, pending.pdsUrl, pending.issuer)
        val session = OAuthSession(
            accountDid = subject,
            handle = pending.inputIdentifier.takeUnless { it.startsWith("did:") },
            pdsUrl = pending.pdsUrl,
            issuer = pending.issuer,
            tokenEndpoint = pending.tokenEndpoint,
            accessToken = token.getString("access_token"),
            refreshToken = refreshToken,
            scopes = scopes,
            expiresAtEpochMillis = System.currentTimeMillis() +
                token.getLong("expires_in") * 1_000,
            dpopKeyId = pending.dpopKeyId,
            authorizationServerNonce = result.second,
            resourceServerNonce = null,
            flashesActorProfileReady = null,
        )
        store.saveSession(session)
        store.deletePending()
        return ensureFlashesActorProfile(session)
    }

    fun currentSession(): OAuthSession? = store.loadSession()

    suspend fun validSession(): OAuthSession = refreshMutex.withLock {
        val session = store.loadSession()
            ?: throw OAuthException("Connect an AT Protocol account.")
        if (!session.scopes.containsAll(OAuthConfig.Scopes)) {
            signOut()
            throw OAuthException("Your Presently session needs to be connected again.")
        }
        if (session.expiresAtEpochMillis - System.currentTimeMillis() > 30_000) {
            return@withLock ensureFlashesActorProfile(session)
        }
        ensureFlashesActorProfile(refresh(session))
    }

    suspend fun publishStory(
        jpegData: ByteArray,
        createdAt: Instant,
        recordKey: String,
    ): PublishedStory {
        if (jpegData.size > FlashesStoryContract.MaximumImageBytes) {
            throw StoryPublishingException("This photo is larger than Flashes’ 10 MiB limit.")
        }

        var session = validSession()
        val upload = dpopResourceRequest(
            url = repoEndpoint(session.pdsUrl, "com.atproto.repo.uploadBlob"),
            body = jpegData,
            contentType = FlashesStoryContract.JpegMimeType,
            session = session,
        )
        session = upload.second
        store.saveSession(session)
        if (upload.first.statusCode != 200) {
            throw StoryPublishingException("The photo upload failed. Please try again.")
        }
        val blobObject = runCatching {
            JSONObject(String(upload.first.body, Charsets.UTF_8)).getJSONObject("blob")
        }.getOrElse {
            throw StoryPublishingException("The photo server returned an invalid upload.")
        }
        val reference = blobObject.optJSONObject("ref")
        val blob = ATProtoBlob(
            type = blobObject.optString("\$type", "blob"),
            cid = reference?.optString("\$link").orEmpty(),
            mimeType = blobObject.optString("mimeType"),
            size = blobObject.optInt("size", -1),
        )
        if (blob.cid.isBlank() ||
            blob.mimeType != FlashesStoryContract.JpegMimeType ||
            blob.size != jpegData.size
        ) {
            throw StoryPublishingException("The photo server returned an invalid upload.")
        }

        val record = FlashesStoryRecordFactory.create(blob, createdAt)
        val createBody = JSONObject().apply {
            put("repo", session.accountDid)
            put("collection", FlashesStoryContract.Collection)
            put("rkey", recordKey)
            put("record", record.toJson())
        }.toString().toByteArray(Charsets.UTF_8)
        val create = dpopResourceRequest(
            url = repoEndpoint(session.pdsUrl, "com.atproto.repo.createRecord"),
            body = createBody,
            contentType = "application/json",
            session = session,
        )
        store.saveSession(create.second)
        val expectedUri =
            "at://${session.accountDid}/${FlashesStoryContract.Collection}/$recordKey"
        if (isError(create.first, "RecordAlreadyExists")) {
            return PublishedStory(expectedUri, null)
        }
        if (create.first.statusCode != 200) {
            throw StoryPublishingException("The story could not be posted. Please try again.")
        }
        val response = runCatching {
            JSONObject(String(create.first.body, Charsets.UTF_8))
        }.getOrElse {
            throw StoryPublishingException("The photo server returned an invalid story.")
        }
        val uri = response.optString("uri")
        val cid = response.optString("cid")
        if (uri != expectedUri || cid.isBlank()) {
            throw StoryPublishingException("The photo server returned an invalid story.")
        }
        return PublishedStory(uri, cid)
    }

    fun signOut() {
        store.loadSession()?.let { runCatching { DPoPKey.load(it.dpopKeyId).delete() } }
        store.loadPending()?.let { runCatching { DPoPKey.load(it.dpopKeyId).delete() } }
        store.deleteSession()
        store.deletePending()
    }

    fun cancelPendingAuthorization() {
        store.loadPending()?.let {
            runCatching { DPoPKey.load(it.dpopKeyId).delete() }
        }
        store.deletePending()
    }

    private suspend fun refresh(session: OAuthSession): OAuthSession {
        val key = DPoPKey.load(session.dpopKeyId)
        val result = dpopFormRequest(
            session.tokenEndpoint,
            listOf(
                "grant_type" to "refresh_token",
                "client_id" to OAuthConfig.ClientId,
                "refresh_token" to session.refreshToken,
            ),
            key,
            session.authorizationServerNonce,
        )
        if (result.first.statusCode != 200) {
            signOut()
            throw OAuthException("Your Presently session expired. Connect it again.")
        }
        val token = JSONObject(String(result.first.body, Charsets.UTF_8))
        val scopes = token.optString("scope").split(" ").filter(String::isNotBlank)
        if (!token.optString("token_type").equals("DPoP", ignoreCase = true) ||
            token.optString("sub") != session.accountDid ||
            !scopes.containsAll(OAuthConfig.Scopes) ||
            token.optString("refresh_token").isEmpty()
        ) {
            signOut()
            throw OAuthException("Your Presently session is no longer valid.")
        }
        val refreshed = session.copy(
            accessToken = token.getString("access_token"),
            refreshToken = token.getString("refresh_token"),
            scopes = scopes,
            expiresAtEpochMillis = System.currentTimeMillis() +
                token.getLong("expires_in") * 1_000,
            authorizationServerNonce = result.second,
        )
        store.saveSession(refreshed)
        return refreshed
    }

    private suspend fun ensureFlashesActorProfile(session: OAuthSession): OAuthSession {
        if (session.flashesActorProfileReady == true) return session
        val lookupUrl = repoEndpoint(
            session.pdsUrl,
            "com.atproto.repo.getRecord",
            listOf(
                "repo" to session.accountDid,
                "collection" to FlashesActorProfileContract.Collection,
                "rkey" to FlashesActorProfileContract.RecordKey,
            ),
        )
        val lookup = http.get(lookupUrl, mapOf("Accept" to "application/json"))
        if (lookup.statusCode == 200) {
            return session.copy(flashesActorProfileReady = true).also(store::saveSession)
        }
        if (!isError(lookup, "RecordNotFound")) {
            throw OAuthException("Presently couldn’t prepare this account for stories.")
        }

        val body = JSONObject().apply {
            put("repo", session.accountDid)
            put("collection", FlashesActorProfileContract.Collection)
            put("rkey", FlashesActorProfileContract.RecordKey)
            put("record", FlashesActorProfileFactory.create(Instant.now()))
        }.toString().toByteArray(Charsets.UTF_8)
        val create = dpopResourceRequest(
            repoEndpoint(session.pdsUrl, "com.atproto.repo.createRecord"),
            body,
            "application/json",
            session,
        )
        if (create.first.statusCode != 200 &&
            !isError(create.first, "RecordAlreadyExists")
        ) {
            throw OAuthException("Presently couldn’t prepare this account for stories.")
        }
        return create.second.copy(flashesActorProfileReady = true).also(store::saveSession)
    }

    private suspend fun dpopResourceRequest(
        url: String,
        body: ByteArray,
        contentType: String,
        session: OAuthSession,
    ): Pair<OAuthHttpResponse, OAuthSession> {
        val key = DPoPKey.load(session.dpopKeyId)
        var nonce = session.resourceServerNonce
        var updatedSession = session
        repeat(2) { attempt ->
            val response = http.post(
                url,
                body,
                contentType,
                mapOf(
                    "Authorization" to "DPoP ${session.accessToken}",
                    "DPoP" to key.proof(
                        method = "POST",
                        url = url,
                        nonce = nonce,
                        accessToken = session.accessToken,
                    ),
                ),
            )
            response.dpopNonce?.let {
                nonce = it
                updatedSession = updatedSession.copy(resourceServerNonce = it)
            }
            if (attempt == 0 && response.statusCode in listOf(400, 401) &&
                isError(response, "use_dpop_nonce")
            ) {
                return@repeat
            }
            return response to updatedSession
        }
        throw OAuthException("The photo server did not accept its security nonce.")
    }

    private fun repoEndpoint(
        pdsUrl: String,
        method: String,
        queryItems: List<Pair<String, String>> = emptyList(),
    ): String {
        val base = "${pdsUrl.trimEnd('/')}/xrpc/$method"
        if (queryItems.isEmpty()) return base
        return "$base?" + queryItems.joinToString("&") { (name, value) ->
            "${name.formEncode()}=${value.formEncode()}"
        }
    }

    private fun isError(response: OAuthHttpResponse, expected: String): Boolean =
        runCatching {
            JSONObject(String(response.body, Charsets.UTF_8)).optString("error") == expected
        }.getOrDefault(false)

    private suspend fun dpopFormRequest(
        url: String,
        fields: List<Pair<String, String>>,
        key: DPoPKey,
        startingNonce: String?,
    ): Pair<OAuthHttpResponse, String> {
        var nonce = startingNonce
        repeat(2) { attempt ->
            val response = http.postForm(
                url,
                fields,
                mapOf("DPoP" to key.proof("POST", url, nonce)),
            )
            val responseNonce = response.dpopNonce
                ?: throw OAuthException("The server omitted its required DPoP nonce.")
            nonce = responseNonce
            val error = runCatching {
                JSONObject(String(response.body, Charsets.UTF_8)).optString("error")
            }.getOrNull()
            if (attempt == 0 && response.statusCode in listOf(400, 401) &&
                error == "use_dpop_nonce"
            ) {
                return@repeat
            }
            return response to responseNonce
        }
        throw OAuthException("The server did not accept its DPoP nonce.")
    }

    private fun validateCallback(url: String, pending: PendingAuthorization): String {
        val uri = URI(url)
        val expected = URI(OAuthConfig.RedirectUri)
        if (uri.scheme != expected.scheme || uri.host != expected.host ||
            uri.path != expected.path
        ) {
            throw OAuthException("The OAuth callback URI was invalid.")
        }
        val values = mutableMapOf<String, String>()
        uri.rawQuery.orEmpty().split("&")
            .filter(String::isNotEmpty)
            .forEach {
                val parts = it.split("=", limit = 2)
                val name = parts[0].formDecode()
                if (values.containsKey(name)) {
                    throw OAuthException("The OAuth callback included duplicate parameters.")
                }
                values[name] = parts.getOrElse(1) { "" }.formDecode()
            }
        if (values["state"] != pending.state) {
            throw OAuthException("The OAuth callback state did not match.")
        }
        if (values["iss"] != pending.issuer) {
            throw OAuthException("The OAuth callback came from an unexpected server.")
        }
        values["error"]?.let {
            throw OAuthException(values["error_description"] ?: it)
        }
        return values["code"]?.takeIf(String::isNotEmpty)
            ?: throw OAuthException("The OAuth callback did not include a code.")
    }

    private fun randomToken(size: Int = 32): String =
        ByteArray(size).also(SecureRandom()::nextBytes).base64Url()

    private fun ByteArray.base64Url(): String = Base64.encodeToString(
        this,
        Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING,
    )

    private fun String.formEncode() = URLEncoder.encode(this, Charsets.UTF_8.name())
    private fun String.formDecode() = URLDecoder.decode(this, Charsets.UTF_8.name())
}

File: apps/ios/Presently/OAuth/ATProtoOAuthClient.swift
import CryptoKit
import Foundation
import Security

actor ATProtoOAuthClient {
    private let store: OAuthSecureStore
    private let http: OAuthHTTPClient
    private let discovery: OAuthDiscovery

    init(
        store: OAuthSecureStore = OAuthSecureStore(),
        http: OAuthHTTPClient = OAuthHTTPClient()
    ) {
        self.store = store
        self.http = http
        discovery = OAuthDiscovery(http: http)
    }

    func prepareAuthorization(identifier: String) async throws -> URL {
        let clientMetadata: OAuthClientMetadata = try await http.getJSON(
            PresentlyOAuthConfiguration.clientID
        )
        guard PresentlyOAuthConfiguration
            .publishedMetadataSupportsRequiredScopes(clientMetadata.scope) else {
            throw OAuthValidationError.clientMetadataOutOfDate
        }

        let result = try await discovery.discover(identifier: identifier)
        let key = try DPoPKey(store: store)
        do {
            let state = randomToken()
            let verifier = randomToken(byteCount: 48)
            let challenge = Data(
                SHA256.hash(data: Data(verifier.utf8))
            ).base64URLEncodedString()

            let fields = [
                URLQueryItem(
                    name: "client_id",
                    value: PresentlyOAuthConfiguration.clientID.absoluteString
                ),
                URLQueryItem(name: "response_type", value: "code"),
                URLQueryItem(
                    name: "redirect_uri",
                    value: PresentlyOAuthConfiguration.redirectURI.absoluteString
                ),
                URLQueryItem(
                    name: "scope",
                    value: PresentlyOAuthConfiguration.scope
                ),
                URLQueryItem(name: "state", value: state),
                URLQueryItem(name: "code_challenge", value: challenge),
                URLQueryItem(name: "code_challenge_method", value: "S256"),
                URLQueryItem(name: "login_hint", value: result.inputIdentifier),
            ]
            let par = try await dpopFormRequest(
                url: result.server.pushedAuthorizationRequestEndpoint,
                fields: fields,
                key: key,
                nonce: nil
            )
            guard par.response.statusCode == 200 || par.response.statusCode == 201,
                  let object = try JSONSerialization.jsonObject(
                    with: par.response.data
                  ) as? [String: Any],
                  let requestURI = object["request_uri"] as? String,
                  !requestURI.isEmpty else {
                throw OAuthValidationError.unexpectedResponse(
                    par.response.statusCode
                )
            }

            let pending = PendingAuthorization(
                inputIdentifier: result.inputIdentifier,
                expectedDID: result.did,
                pdsURL: result.pdsURL,
                issuer: result.server.issuer,
                authorizationEndpoint: result.server.authorizationEndpoint,
                tokenEndpoint: result.server.tokenEndpoint,
                state: state,
                codeVerifier: verifier,
                dpopKeyID: key.id,
                authorizationServerNonce: par.nonce
            )
            try store.savePendingAuthorization(pending)

            var components = URLComponents(
                url: result.server.authorizationEndpoint,
                resolvingAgainstBaseURL: false
            )
            components?.queryItems = [
                URLQueryItem(
                    name: "client_id",
                    value: PresentlyOAuthConfiguration.clientID.absoluteString
                ),
                URLQueryItem(name: "request_uri", value: requestURI),
            ]
            guard let authorizationURL = components?.url else {
                throw OAuthValidationError.invalidAuthorizationServerMetadata
            }
            return authorizationURL
        } catch {
            try? store.deleteDPoPPrivateKey(id: key.id)
            throw error
        }
    }

    func completeAuthorization(callbackURL: URL) async throws -> OAuthSession {
        guard var pending = try store.loadPendingAuthorization() else {
            throw OAuthValidationError.invalidCallback
        }
        let callback = try validateCallback(callbackURL, pending: pending)
        let key = try DPoPKey(id: pending.dpopKeyID, store: store)
        let fields = [
            URLQueryItem(name: "grant_type", value: "authorization_code"),
            URLQueryItem(
                name: "client_id",
                value: PresentlyOAuthConfiguration.clientID.absoluteString
            ),
            URLQueryItem(
                name: "redirect_uri",
                value: PresentlyOAuthConfiguration.redirectURI.absoluteString
            ),
            URLQueryItem(name: "code", value: callback.code),
            URLQueryItem(name: "code_verifier", value: pending.codeVerifier),
        ]
        let tokenResult = try await dpopFormRequest(
            url: pending.tokenEndpoint,
            fields: fields,
            key: key,
            nonce: pending.authorizationServerNonce
        )
        pending.authorizationServerNonce = tokenResult.nonce
        try store.savePendingAuthorization(pending)

        guard tokenResult.response.statusCode == 200,
              let token = try? JSONDecoder().decode(
                TokenResponse.self,
                from: tokenResult.response.data
              ),
              token.tokenType.caseInsensitiveCompare("DPoP") == .orderedSame,
              let refreshToken = token.refreshToken,
              !refreshToken.isEmpty else {
            throw OAuthValidationError.invalidTokenResponse
        }
        guard token.subject == pending.expectedDID else {
            throw OAuthValidationError.accountMismatch
        }
        let grantedScopes = token.scope.split(separator: " ").map(String.init)
        guard Set(PresentlyOAuthConfiguration.scopes)
            .isSubset(of: Set(grantedScopes)) else {
            throw OAuthValidationError.insufficientScope
        }

        try await discovery.validateAccount(
            token.subject,
            pdsURL: pending.pdsURL,
            issuer: pending.issuer
        )

        let session = OAuthSession(
            accountDID: token.subject,
            handle: pending.inputIdentifier.hasPrefix("did:")
                ? nil
                : pending.inputIdentifier,
            pdsURL: pending.pdsURL,
            issuer: pending.issuer,
            tokenEndpoint: pending.tokenEndpoint,
            accessToken: token.accessToken,
            refreshToken: refreshToken,
            scopes: grantedScopes,
            expiresAt: Date().addingTimeInterval(TimeInterval(token.expiresIn)),
            dpopKeyID: pending.dpopKeyID,
            authorizationServerNonce: tokenResult.nonce,
            resourceServerNonce: nil,
            flashesActorProfileReady: nil
        )
        try store.saveSession(session)
        try store.deletePendingAuthorization()
        return try await ensureFlashesActorProfile(for: session)
    }

    func currentSession() throws -> OAuthSession? {
        try store.loadSession()
    }

    func validSession() async throws -> OAuthSession {
        guard let session = try store.loadSession() else {
            throw OAuthValidationError.sessionExpired
        }
        guard Set(PresentlyOAuthConfiguration.scopes)
            .isSubset(of: Set(session.scopes)) else {
            try? signOut()
            throw OAuthValidationError.insufficientScope
        }
        let activeSession = session.expiresAt.timeIntervalSinceNow > 30
            ? session
            : try await refresh(session)
        return try await ensureFlashesActorProfile(for: activeSession)
    }

    func signOut() throws {
        if let session = try store.loadSession() {
            try? store.deleteDPoPPrivateKey(id: session.dpopKeyID)
        }
        if let pending = try store.loadPendingAuthorization() {
            try? store.deleteDPoPPrivateKey(id: pending.dpopKeyID)
        }
        try store.deleteSession()
        try store.deletePendingAuthorization()
    }

    func publishStory(
        jpegData: Data,
        createdAt: Date,
        recordKey: String
    ) async throws -> PublishedStory {
        guard jpegData.count <= FlashesStoryContract.maximumImageBytes else {
            throw StoryContractError.imageTooLarge(jpegData.count)
        }

        var session = try await validSession()
        let uploadURL = try repoEndpoint(
            session.pdsURL,
            method: "com.atproto.repo.uploadBlob"
        )
        let uploadResult = try await dpopResourceRequest(
            url: uploadURL,
            body: jpegData,
            contentType: FlashesStoryContract.jpegMIMEType,
            session: session
        )
        session = uploadResult.session
        try store.saveSession(session)
        guard uploadResult.response.statusCode == 200 else {
            throw StoryContractError.uploadFailed(
                uploadResult.response.statusCode
            )
        }
        guard let uploaded = try? JSONDecoder().decode(
            UploadBlobResponse.self,
            from: uploadResult.response.data
        ), uploaded.blob.mimeType == FlashesStoryContract.jpegMIMEType,
           uploaded.blob.size == jpegData.count else {
            throw StoryContractError.invalidBlobResponse
        }

        let record = try FlashesStoryRecordFactory.make(
            blob: uploaded.blob,
            createdAt: createdAt
        )
        let createURL = try repoEndpoint(
            session.pdsURL,
            method: "com.atproto.repo.createRecord"
        )
        let createBody = try JSONEncoder().encode(
            CreateRecordRequest(
                repo: session.accountDID,
                collection: FlashesStoryContract.collection,
                recordKey: recordKey,
                record: record
            )
        )
        let createResult = try await dpopResourceJSONRequest(
            url: createURL,
            body: createBody,
            session: session
        )
        try store.saveSession(createResult.session)

        if Self.isRecordAlreadyExists(createResult.response) {
            return PublishedStory(
                uri: "at://\(session.accountDID)/\(FlashesStoryContract.collection)/\(recordKey)",
                cid: nil
            )
        }
        guard createResult.response.statusCode == 200 else {
            throw StoryContractError.createRecordFailed(
                createResult.response.statusCode
            )
        }
        guard let response = try? JSONDecoder().decode(
            CreateRecordResponse.self,
            from: createResult.response.data
        ), response.uri ==
            "at://\(session.accountDID)/\(FlashesStoryContract.collection)/\(recordKey)",
           !response.cid.isEmpty else {
            throw StoryContractError.invalidCreateRecordResponse
        }
        return PublishedStory(uri: response.uri, cid: response.cid)
    }

    private func refresh(_ session: OAuthSession) async throws -> OAuthSession {
        let key = try DPoPKey(id: session.dpopKeyID, store: store)
        let result = try await dpopFormRequest(
            url: session.tokenEndpoint,
            fields: [
                URLQueryItem(name: "grant_type", value: "refresh_token"),
                URLQueryItem(
                    name: "client_id",
                    value: PresentlyOAuthConfiguration.clientID.absoluteString
                ),
                URLQueryItem(name: "refresh_token", value: session.refreshToken),
            ],
            key: key,
            nonce: session.authorizationServerNonce
        )
        guard result.response.statusCode == 200,
              let token = try? JSONDecoder().decode(
                TokenResponse.self,
                from: result.response.data
              ),
              token.tokenType.caseInsensitiveCompare("DPoP") == .orderedSame,
              token.subject == session.accountDID,
              let refreshToken = token.refreshToken else {
            try? signOut()
            throw OAuthValidationError.sessionExpired
        }
        let scopes = token.scope.split(separator: " ").map(String.init)
        guard Set(PresentlyOAuthConfiguration.scopes).isSubset(of: Set(scopes)) else {
            try? signOut()
            throw OAuthValidationError.insufficientScope
        }
        let refreshed = OAuthSession(
            accountDID: session.accountDID,
            handle: session.handle,
            pdsURL: session.pdsURL,
            issuer: session.issuer,
            tokenEndpoint: session.tokenEndpoint,
            accessToken: token.accessToken,
            refreshToken: refreshToken,
            scopes: scopes,
            expiresAt: Date().addingTimeInterval(TimeInterval(token.expiresIn)),
            dpopKeyID: session.dpopKeyID,
            authorizationServerNonce: result.nonce,
            resourceServerNonce: session.resourceServerNonce,
            flashesActorProfileReady: session.flashesActorProfileReady
        )
        try store.saveSession(refreshed)
        return refreshed
    }

    private func ensureFlashesActorProfile(
        for session: OAuthSession
    ) async throws -> OAuthSession {
        guard session.flashesActorProfileReady != true else {
            return session
        }

        let lookupURL = try repoEndpoint(
            session.pdsURL,
            method: "com.atproto.repo.getRecord",
            queryItems: [
                URLQueryItem(name: "repo", value: session.accountDID),
                URLQueryItem(
                    name: "collection",
                    value: FlashesActorProfileContract.collection
                ),
                URLQueryItem(
                    name: "rkey",
                    value: FlashesActorProfileContract.recordKey
                ),
            ]
        )
        let lookupResponse = try await http.get(
            lookupURL,
            headers: ["Accept": "application/json"]
        )

        if lookupResponse.statusCode == 200 {
            return try markFlashesActorProfileReady(session)
        }
        guard Self.isRecordNotFound(lookupResponse) else {
            throw OAuthValidationError.unexpectedResponse(
                lookupResponse.statusCode
            )
        }

        let createURL = try repoEndpoint(
            session.pdsURL,
            method: "com.atproto.repo.createRecord"
        )
        let body = try JSONEncoder().encode(
            CreateRecordRequest(
                repo: session.accountDID,
                collection: FlashesActorProfileContract.collection,
                recordKey: FlashesActorProfileContract.recordKey,
                record: FlashesActorProfileFactory.make(createdAt: Date())
            )
        )
        let result = try await dpopResourceJSONRequest(
            url: createURL,
            body: body,
            session: session
        )
        guard result.response.statusCode == 200
                || Self.isRecordAlreadyExists(result.response) else {
            throw OAuthValidationError.unexpectedResponse(
                result.response.statusCode
            )
        }

        var readySession = result.session
        readySession.flashesActorProfileReady = true
        try store.saveSession(readySession)
        return readySession
    }

    private func markFlashesActorProfileReady(
        _ session: OAuthSession
    ) throws -> OAuthSession {
        var readySession = session
        readySession.flashesActorProfileReady = true
        try store.saveSession(readySession)
        return readySession
    }

    private func dpopResourceJSONRequest(
        url: URL,
        body: Data,
        session: OAuthSession
    ) async throws -> (response: OAuthHTTPResponse, session: OAuthSession) {
        try await dpopResourceRequest(
            url: url,
            body: body,
            contentType: "application/json",
            session: session
        )
    }

    private func dpopResourceRequest(
        url: URL,
        body: Data,
        contentType: String,
        session: OAuthSession
    ) async throws -> (response: OAuthHTTPResponse, session: OAuthSession) {
        let key = try DPoPKey(id: session.dpopKeyID, store: store)
        var activeNonce = session.resourceServerNonce
        var updatedSession = session

        for attempt in 0..<2 {
            let proof = try key.proof(
                method: "POST",
                url: url,
                nonce: activeNonce,
                accessToken: session.accessToken
            )
            let response = try await http.post(
                url,
                body: body,
                contentType: contentType,
                headers: [
                    "Authorization": "DPoP \(session.accessToken)",
                    "DPoP": proof,
                ]
            )
            if let responseNonce = response.dpopNonce {
                activeNonce = responseNonce
                updatedSession.resourceServerNonce = responseNonce
            }
            if attempt == 0, responseRequiresDPoPNonce(response) {
                continue
            }
            return (response, updatedSession)
        }
        throw OAuthValidationError.dpopNonceMissing
    }

    private func repoEndpoint(
        _ pdsURL: URL,
        method: String,
        queryItems: [URLQueryItem] = []
    ) throws -> URL {
        let endpoint = pdsURL
            .appendingPathComponent("xrpc")
            .appendingPathComponent(method)
        guard var components = URLComponents(
            url: endpoint,
            resolvingAgainstBaseURL: false
        ) else {
            throw OAuthValidationError.invalidProtectedResourceMetadata
        }
        components.queryItems = queryItems.isEmpty ? nil : queryItems
        guard let url = components.url else {
            throw OAuthValidationError.invalidProtectedResourceMetadata
        }
        return url
    }

    static func isRecordNotFound(_ response: OAuthHTTPResponse) -> Bool {
        response.statusCode == 400
            && responseError(response) == "RecordNotFound"
    }

    static func isRecordAlreadyExists(_ response: OAuthHTTPResponse) -> Bool {
        response.statusCode == 400
            && responseError(response) == "RecordAlreadyExists"
    }

    private static func responseError(
        _ response: OAuthHTTPResponse
    ) -> String? {
        let object = try? JSONSerialization.jsonObject(
            with: response.data
        ) as? [String: Any]
        return object?["error"] as? String
    }

    private func dpopFormRequest(
        url: URL,
        fields: [URLQueryItem],
        key: DPoPKey,
        nonce: String?
    ) async throws -> (response: OAuthHTTPResponse, nonce: String) {
        var activeNonce = nonce
        for attempt in 0..<2 {
            let proof = try key.proof(
                method: "POST",
                url: url,
                nonce: activeNonce
            )
            let response = try await http.postForm(
                url,
                fields: fields,
                headers: ["DPoP": proof]
            )
            guard let responseNonce = response.dpopNonce else {
                throw OAuthValidationError.dpopNonceMissing
            }
            activeNonce = responseNonce
            if attempt == 0, responseRequiresDPoPNonce(response) {
                continue
            }
            return (response, responseNonce)
        }
        throw OAuthValidationError.dpopNonceMissing
    }

    private func responseRequiresDPoPNonce(_ response: OAuthHTTPResponse) -> Bool {
        guard response.statusCode == 400 || response.statusCode == 401,
              let object = try? JSONSerialization.jsonObject(
                with: response.data
              ) as? [String: Any] else {
            return false
        }
        return object["error"] as? String == "use_dpop_nonce"
    }

    private func validateCallback(
        _ url: URL,
        pending: PendingAuthorization
    ) throws -> (code: String, issuer: String) {
        guard url.scheme == PresentlyOAuthConfiguration.redirectURI.scheme,
              url.host == PresentlyOAuthConfiguration.redirectURI.host,
              url.path == PresentlyOAuthConfiguration.redirectURI.path,
              let components = URLComponents(
                url: url,
                resolvingAgainstBaseURL: false
              ) else {
            throw OAuthValidationError.invalidCallback
        }
        var values: [String: String] = [:]
        for item in components.queryItems ?? [] {
            guard values[item.name] == nil else {
                throw OAuthValidationError.invalidCallback
            }
            values[item.name] = item.value ?? ""
        }
        guard values["state"] == pending.state else {
            throw OAuthValidationError.callbackStateMismatch
        }
        guard values["iss"] == pending.issuer.absoluteString else {
            throw OAuthValidationError.callbackIssuerMismatch
        }
        if let error = values["error"] {
            throw OAuthValidationError.authorizationDenied(
                values["error_description"] ?? error
            )
        }
        guard let code = values["code"], !code.isEmpty else {
            throw OAuthValidationError.invalidCallback
        }
        return (code, values["iss"]!)
    }

    private func randomToken(byteCount: Int = 32) -> String {
        var bytes = [UInt8](repeating: 0, count: byteCount)
        _ = SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes)
        return Data(bytes).base64URLEncodedString()
    }
}
